docs: add documentation for the --experimental-include-runtime flag (draft) (#15750)
cloudflare/cloudflare-docs: content/workers/languages/typescript/_index.md

TypeScript is a first-class language on Cloudflare Workers. Cloudflare publishes type definitions to [GitHub](https://github.com/cloudflare/workers-types) and [npm](https://www.npmjs.com/package/@cloudflare/workers-types) (`npm install -D @cloudflare/workers-types`). All APIs provided in Workers are fully typed, and type definitions are generated directly from [workerd](https://github.com/cloudflare/workerd), the open-source Workers runtime.

### Dynamic Runtime Types (Experimental)

> Your runtime APIs are dynamic - their exact configuration depends on:

Reviewer: Think this should explain why to do this

TypeScript is a first-class language on Cloudflare Workers. Cloudflare publishes type definitions to [GitHub](https://github.com/cloudflare/workers-types) and [npm](https://www.npmjs.com/package/@cloudflare/workers-types) (`npm install -D @cloudflare/workers-types`). All APIs provided in Workers are fully typed, and type definitions are generated directly from [workerd](https://github.com/cloudflare/workerd), the open-source Workers runtime.

> ### Dynamic Runtime Types (Experimental)

Reviewer: ```suggestion
### Generate Types that match your Worker's configuration (Experimental)
```

Something like this to be more concrete?

> There is a possibility that the types defined in `@cloudflare/workers-types` might not exactly match your project's runtime APIs. This is because runtime APIs are dynamic, and their exact configuration depends on:

Reviewer: ```suggestion
Cloudflare continuously improves [workerd](https://github.com/cloudflare/workerd), the open-source Workers runtime.

Some updates change the TypeScript types that APIs return. For example, the [`urlsearchparams_delete_has_value_arg`](/workers/configuration/compatibility-dates/#urlsearchparams-delete-and-has-value-argument) compatibility flag adds optional arguments to some methods, in order to support new additions to the WHATWG URL standard API.

This means the correct TypeScript types for your Worker depend on:
```

> 1. The `compatibility_date` defined in your `wrangler.toml`.

Reviewer: ```suggestion
1. Your Worker's [Compatibility Date](/workers/configuration/compatibility-dates/)
```

> 2. Any `compatibility_flags` defined in your `wrangler.toml`.

Reviewer: ```suggestion
2. Your Worker's [Compatibility Flags](/workers/configuration/compatibility-dates/#compatibility-flags)
```

> For example, if you have `compatibility_flags = ["nodejs_als"]` in your `wrangler.toml`, then the runtime will allow you to use the [AsyncLocalStorage](https://nodejs.org/api/async_context.html#class-asynclocalstorage) class in your worker code, but you will not see this reflected in the type definitions in `@cloudflare/workers-types`.

Reviewer: To me this implies that if I set `compatibility_flags = ["nodejs_compat"]` — then I'll get types back for Node.js APIs (because ALS is a subset of this)

But that's not something that we support (we should say this explicitly!)

Is there maybe an example of a different compat flag that we could use here, for clarity?

And also...is there an explicit reason why `nodejs_compat` doesn't yet work but `nodejs_als` does?

Reviewer: ```suggestion
For example, if you have `compatibility_flags = ["nodejs_als"]` in your `wrangler.toml`, then the [AsyncLocalStorage](https://nodejs.org/api/async_context.html#class-asynclocalstorage) API can be used in your Worker, but you will not see this reflected in the type definitions in `@cloudflare/workers-types`.
```

> In order to solve this issue, and to ensure that your type definitions are always up-to-date with your compatibility settings, you can dynamically generate the runtime types (as of `wrangler 3.66.1`):

Reviewer: ```suggestion
To ensure that the type definitions you rely on match runtime behavior, as of `wrangler 3.66.1`, you can run a Wrangler command to generate TypeScript types based on your compatibility date and flags:
```

```bash
wrangler types --experimental-runtime

# You could also use the alias

> wrangler types --x-runtime

Reviewer: I would suggest picking one and sticking with it in context of these docs. If the Wrangler docs on commands spell out both, that seems fine.

Reviewer 2: I'm also not that convinced that `--x-runtime` is that much shorter/better than `--experimental-runtime` I'd personally just go with the extended version 🤔

Author: This is here in order to maintain consistency with other experimental flags in wrangler. As far as I know, they all have (or should have) an `--x` alias.

Reviewer 2: ah ok sorry I didn't know/wasn't sure about that 🙂 

if it's an established pattern then yeah I agree that we should just follow it 👍

```

> This will generate a `d.ts` file and (by default) save it to `.wrangler/types/runtime.d.ts`. You will be prompted in the command's output to add that file to your `tsconfig.json`'s `compilerOptions.types` array.

Reviewer: ```suggestion
This will generate a `d.ts` file and (by default) save it to `.wrangler/types/runtime.d.ts`. After running the command, you must add the path of the generated runtime types file to the [`compilerOptions.types` field](https://www.typescriptlang.org/tsconfig/#types) in your project's [tsconfig.json](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html) file.
```

If you would like to commit the file to git, you can provide a custom path. Here, for instance, the `runtime.d.ts` file will be saved to the root of your project:

```bash
wrangler types --x-include-runtime="./runtime.d.ts"
```

> Note that the `--x-runtime` command replaces the need for the `@cloudflare/workers-types` package, so that package should be uninstalled to avoid any potential for conflict.

Reviewer: ```suggestion
Note that the `--x-runtime` command replaces the need for the `@cloudflare/workers-types` package, you should uninstall that package to avoid conflicts.
```

> ### Type Checking (Experimental)

Reviewer: Not reviewing below this since removing for now
